File: grind-75/matrix/zeroOneMatrix.py
from collections import deque
from typing import List


class Solution:
    @classmethod
    def updateMatrix(self, mat: List[List[int]]) -> List[List[int]]:
        q = deque()
        visited = set()
        directions = [[1, 0], [-1, 0], [0, 1], [0, -1]]

        for r in range(len(mat)):
            for c in range(len(mat[0])):
                if mat[r][c] == 0:
                    visited.add((r, c))
                    q.append((r, c))

        while q:
            r, c = q.popleft()
            for x, y in directions:
                nr = r + x
                nc = c + y
                if (
                    0 <= nr < len(mat)
                    and 0 <= nc < len(mat[0])
                    and (nr, nc) not in visited
                ):
                    visited.add((nr, nc))
                    q.append((nr, nc))
                    mat[nr][nc] = mat[r][c] + 1
        return mat


mat = [[0, 0, 0], [0, 1, 0], [0, 0, 0]]
output = [[0, 0, 0], [0, 1, 0], [0, 0, 0]]
res = Solution.updateMatrix(mat) == output
print(res)

mat = [[0, 0, 0], [0, 1, 0], [1, 1, 1]]
output = [[0, 0, 0], [0, 1, 0], [1, 2, 1]]
res = Solution.updateMatrix(mat) == output
print(res)

File: projects/grind-75/typescript/matrix/zero-one-matrix.ts
export default function updateMatrix(mat: number[][]): number[][] {
    const q: [number, number][] = [];
    const visited = new Set<string>();
    const directions = [[1, 0], [-1, 0], [0, 1], [0, -1]];

    for (let r = 0; r < mat.length; r++) {
        for (let c = 0; c < mat[0].length; c++) {
            if (mat[r][c] === 0) {
                visited.add(`${r},${c}`);
                q.push([r, c]);
            }
        }
    }

    while (q.length > 0) {
        const [r, c] = q.shift()!;
        for (const [x, y] of directions) {
            const nr = r + x;
            const nc = c + y;
            if (
                nr >= 0 && nr < mat.length &&
                nc >= 0 && nc < mat[0].length &&
                !visited.has(`${nr},${nc}`)
            ) {
                visited.add(`${nr},${nc}`);
                q.push([nr, nc]);
                mat[nr][nc] = mat[r][c] + 1;
            }
        }
    }

    return mat;
};